### DISCUSSION TITLE
Date field transmitted as json not converted to Date

### DISCUSSION THREAD

**Discussion by arnaud16571542** [loopbackio/loopback-next#7499] [Q&A]
Hello everyone,

I have a stupid question i'm struggle with.

My Nuxt/Axios front-end do a JSON post request to my loopback backend : using a DTO model object which contain a Date field.
```
@model()
export class MyDTO extends Model {

  @property({
    type: 'date',
    required: false,
  })
  startDate?: Date;

  constructor(data?: Partial<MyDTO>) {
    super(data);
  }
}
```

Axios convert all the objects to JSON. The date field is converted to something like '2021-05-20T09:54:47.849Z'.

```
The Request Payload is look like : 
{...,"startDate":"2021-05-20T22:00:00.000Z" }
```

My problem is : this date, declared on loopback as type:'date'/javascript as Date, is filled as _string_ in the DTO object and not as Date during parsing. This cause cast problem lately in the controller.

I don't have this problem for boolean value. So, for me, this doesn't seem the correct behaviours for Date fields. Probably i miss a configuration, but I can't find it :(

```
  @post('/myfunction', {
    responses: {
      '200': {
        content: { 'application/json': { schema: getModelSchemaRef(Result) } },
      },
    },
  })
  async create(
    @requestBody() myDTO: MyDTO
  ): Promise<Result> {
   console.log(typeof myDTO.startDate) // <= Result is 'string'. I was expecting Date.. and typescript too !
}
```
Any idea ? Thank you for your help !

**Comment by marioestradarosa** [comment]
@arnaud16571542 ,  Date type do not exist at run time, so when running this piece of code it is not showing Date (since it does not exists, rather it shows its corresponding type `string | object ??`), this is not loopback4 behavior but javascript/typescript.  

What you can do at compile-time is to use type-guard functions as explained here [using-type-predicates](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) .  

```ts
function isValidDate(d) {
  return d instanceof Date && !isNaN(d);
}
```

Be asure however, that loopback4 will check the content upon request, that's the function for the parser 
and validations using `ajv`  json schema validator annotated by `@requestBody`.

**Reply by arnaud16571542** [reply]
Thank you for your reply !

I checked again this page (https://loopback.io/doc/en/lb4/Parsing-requests.html), and from my understanding (with your explanation & the page) :
In my exemple:    @requestBody() myDTO: MyDTO

@requestBody use MyDTO only to find the corresponding Model, get & apply JSON validation rules from properties, and it creates a Object filled with string/boolean/other basic types data.
- Complex data object, even like Date(), stay in json format and copied into MyDTO. 

The documentation is not clear about body parameters, but now I understand I must (and it is normal) change my Date as string (typescript data type in DTO Model, but keeping loopback 'type' as 'date'), and do my casting manually

I'm somewhat surprised, as Date data type seem well managed by loopback, so I expected an automatic casting, but it is not a big deal :).

Finally, i have done things like that :
```
@model()
export class MyDTO extends Model {
  @property({ type: 'date', required:true })
  startDate: any;  // ( or string|Date)

  constructor(data?: Partial<MyDTO>) { super(data);}
}
```
```
  @post('/myfunction', {
    responses: {
      '200': {  content: { 'application/json': { schema: getModelSchemaRef(Result) } }  },
    },
  })
  async create(
    @requestBody() myDTO: MyDTO
  ): Promise<Result> {
   myDTO.startDate = parseISO(myDTO.startDate)
   ....
}
```
I must precise my DTO object is only for exchange between my frontend & backend, and is not stored.
